x/stake/keeper.go

```diff
@@ -33,463 +30,647 @@ func NewKeeper(cdc *wire.Codec, key sdk.StoreKey, ck bank.Keeper, codespace sdk.
+		}
+		address := iterator.Value()
+		validator, found := k.getValidator(store, address)
+		if !found {
+			panic(fmt.Sprintf("validator record not found for address: %v\n", address))
+		}
 
-	// if the voting power is the same no need to update any of the other indexes
-	if oldFound && oldCandidate.Assets.Equal(candidate.Assets) {
-		return
+		validators[i] = validator
+		i++
 	}
+	iterator.Close()
+	return validators[:i] // trim
+}
 
-	updateHeight := false
-
-	// update the list ordered by voting power
-	if oldFound {
-		if !k.isNewValidator(ctx, store, candidate.Address) {
-			updateHeight = true
+// get the group of bonded validators sorted by power-rank
+func (k Keeper) GetValidatorsByPower(ctx sdk.Context) []Validator {
+	store := ctx.KVStore(k.storeKey)
+	maxValidators := k.GetParams(ctx).MaxValidators
+	validators := make([]Validator, maxValidators)
+	iterator := store.ReverseSubspaceIterator(ValidatorsByPowerKey) // largest to smallest
+	i := 0
+	for {
+		if !iterator.Valid() || i > int(maxValidators-1) {
+			iterator.Close()
+			break
 		}
-		// else already in the validator set - retain the old validator height and counter
-		store.Delete(GetValidatorKey(address, oldCandidate.Assets, oldCandidate.ValidatorBondHeight, oldCandidate.ValidatorBondCounter, k.cdc))
-	} else {
-		updateHeight = true
+		address := iterator.Value()
+		validator, found := k.getValidator(store, address)
+		if !found {
+			panic(fmt.Sprintf("validator record not found for address: %v\n", address))
+		}
+		if validator.Status() == sdk.Bonded {
+			validators[i] = validator
+			i++
+		}
+		iterator.Next()
 	}
+	return validators[:i] // trim
+}
 
-	if updateHeight {
-		// wasn't a candidate or wasn't in the validator set, update the validator block height and counter
-		candidate.ValidatorBondHeight = ctx.BlockHeight()
-		counter := k.getCounter(ctx)
-		candidate.ValidatorBondCounter = counter
-		k.setCounter(ctx, counter+1)
-	}
+//_________________________________________________________________________
+// Accumulated updates to the active/bonded validator set for tendermint
+
+// get the most recently updated validators
+func (k Keeper) getTendermintUpdates(ctx sdk.Context) (updates []abci.Validator) {
+	store := ctx.KVStore(k.storeKey)
 
-	// update the candidate record
-	bz, err = k.cdc.MarshalBinary(candidate)
-	if err != nil {
-		panic(err)
+	iterator := store.SubspaceIterator(TendermintUpdatesKey) //smallest to largest
+	for ; iterator.Valid(); iterator.Next() {
+		valBytes := iterator.Value()
+		var val abci.Validator
+		k.cdc.MustUnmarshalBinary(valBytes, &val)
+		updates = append(updates, val)
 	}
-	store.Set(GetCandidateKey(candidate.Address), bz)
+	iterator.Close()
+	return
+}
 
-	// marshal the new validator record
-	validator := candidate.validator()
-	bz, err = k.cdc.MarshalBinary(validator)
-	if err != nil {
-		panic(err)
+// remove all validator update entries after applied to Tendermint
+func (k Keeper) clearTendermintUpdates(ctx sdk.Context) {
```

Comment:
We should add a general `store.DeleteSubspace` helper function (not in this PR).

Reply:
yeah good idear

Reply:
https://github.com/cosmos/cosmos-sdk/issues/1046